### PR TITLE
[WIP] refactoring PlutoUI with Mustache

### PR DESCRIPTION
**Pull request by GiggleLiu**
I rewrote `Buildins.jl` with the new macro `@widget` and the template language `Mustache.jl`, this is just an experimental version. Some wedgets are not well tested yet.

e.g.
```julia
julia> @widget CheckBox """<input type="checkbox" {{#default}}checked{{/default}}>""" default=false

julia> html(CheckBox())
"<input type=\"checkbox\">"

julia> html(CheckBox(; default=true))
"<input type=\"checkbox\" checked>"
```

The first argument is the type, the second argument is the template and the rest are default values for keyword arguments.

* [Mustache](https://github.com/jverzani/Mustache.jl) is logic-less template "language", it is friendly to frontends.
* The number of code decreased from 397 lines to <100 lines, hence it is much easier to maintain and extend.

### CODE REVIEW

**Comment by GiggleLiu** [comment]
NOTE: The `{{@var}}` grammar requires the merge of https://github.com/jverzani/Mustache.jl/pull/126.

**Comment by dralletje** [comment]
Hey,

Gonna be bold with ya: I don't like it
I do share your interest into making things easier with macros, but I think most of the time it only looks easier.
Yes, the amount of lines has gone down, but at the cost of being readable to people just knowing Julia.
There is a lot of magic going on here, which will (I think) make maintaining this a lot harder.

(Also I don't like Mustache, specifically when we already have Julia's string interpolation)

If I were to go into a "we need less lines" route, I'd go for an option like
https://gist.github.com/dralletje/8d91b4faef661b6fae83263cb499a0d2
Specifically because we are still just working with function composition.

However, I think the way it is now is just fine 😁 

I respect you and our contributions a lot, that's exactly why I am so bold :)

**Comment by GiggleLiu** [comment]
Thanks for your comments, appreciate it. But I am not convinced of the following two points.

> There is a lot of magic going on here, which will (I think) make maintaining this a lot harder.

Maintainability is the problem that I want to solve. Previously, you need to test every widget, check the correctness of the docstrings. Here, you just need to test a macro.
In the previous design, making extension like adding some attributes is hard. As we have discussed in
https://github.com/fonsp/PlutoUI.jl/pull/39
We can specify a list of pairs as `html_attributes`. https://github.com/fonsp/PlutoUI.jl/blob/da9a2d0cfa1e03035c09f2c2087893bb55c6ea5c/src/Resource.jl#L27
But such design pattern does not work for some composite widgets like `Clock`.

> I don't like Mustache, specifically when we already have Julia's string interpolation

This PR care more about HTML/JS/CSS users. **Mustache is logicless, the contrast of powerfulness**. When you allow interpolating a Julia expression, you also open a Pandora's box, then you should make a thorough test.

> If I were to go into a "we need less lines" route, I'd go for an option like
> https://gist.github.com/dralletje/8d91b4faef661b6fae83263cb499a0d2
> Specifically because we are still just working with function composition.

I like this version! The reason behind "we need less line" is, do not repeat the input information. Otherwise, one may run into the maintenance hell.

### After all
I think the division point is you care more for Julia users, while I think styling widgets should be friendly to frontends. Do you think it is better to open a new package to make it more friendly to frontend designers. I also don't like the inconsistent naming with HTML, e.g.
`value <-> default`.

**Comment by GiggleLiu** [comment]
BTW: with this macro, or your `PlutoWidget`. Extending a widget is so much easier. One do not need to learn what is `get` for and what is `show(io, MIME"text/html", obj)` for.

This is important because, I am doing a project that requires styling the widgets.
But I got a lot HTMLs inside the notebook, making the notebook hard to read. This is why I started to writing this macro to make life easier.

File: demo/templates.jl [hunk 1]
```diff
@@ -0,0 +1,70 @@
+include("parser.jl")
+
+@widget FilePicker """<input type='file' accept='{{#accepts}}{{.}}, {{/accepts}}'>""" accepts=[] default=Dict("name" => "", "data" => UInt8[], "type" => "")
```
Comment (jverzani) [inline demo/templates.jl:3]: I'm pretty sure you would want `mt"""..."""` in these definitions so the parsing comes once at compile time and not each time at run time. Should be faster, though perhaps not noticeable.

Comment (GiggleLiu) [reply demo/templates.jl:3]: Perfect, I do not even need to call the parser explicitly. Thanks!

**Comment by dralletje** [comment]
@GiggleLiu you're totally right about it being very complex now to simply add a new UI element.

Maybe we put `PlutoWidget` in, and add your macro as `@MustacheWidget`?
I like having the builtins at least in plain Julia, as these are, I think, more for Julia people to use.
I feel like, if I had to customize the html, I would directly use `@bind x html"<input type=text>"` (or my own templating) rather than of extending `PlutoUI.TextField`.

Then again I think having `@MustacheWidget` would be a great thing for people who do like mustache and some macro magic :) We could make the mustache one more html-centered with, what you said, `value` actually being called `value` :P

---

I'd love to, on the side, have a deep, nerdy discussion about the merits of Mustache 😏 

I hear your love for Mustache, and I once shared it.
However I think that saying "Mustache is logicless, the contrast of powerfulness", though it's beautiful, is incorrect.
I found in my time with Mustache that there very often comes a need for a liiiitle bit of logic. That's is provided by Mustache, with the existence and iteration tags. So already full "logicless" is out of the window. But then as soon as you need a little more than just these tags, I'd start morphing my templates into monsters, just so I could technically still claim to be ~~mostly~~ *logicless*.

Take for example, and this is still a very simple one, the Textfield. I have a pretty hard time understanding what
```julia
"""{{^dims}}<input type="text" value="{{default}}">{{/dims}}
{{@dims}}<textarea {{#dims}}{{#.[1]}}cols="{{.}}"{{/.[1]}} {{#.[2]}}rows="{{.}}"{{/.[2]}}{{/dims}}>{{default}}</textarea>{{/dims}}"""
```
means, though I have no problem reading
```julia
if textfield.dims === nothing
    print(io, """<input type="text" value="$(htmlesc(textfield.default))">""")
else
    print(io, """<textarea cols="$(textfield.dims[1])" rows="$(textfield.dims[2])">$(htmlesc(textfield.default))</textarea>""")
end
```
Even though the last snippet did use print and io and all, which are not ideal.

If, however, you are on a team where someone is designing all the html, and someone is creating all the julia code/values that should go into the html... yeah, then Mustache might be a better way to go than handing raw julia to the designer. However, I wonder how far you can get with some well made macros. I am very curious how a "hypertext literal" variant will look.

**Comment by GiggleLiu** [comment]
> Maybe we put PlutoWidget in, and add your macro as @MustacheWidget?

I added the macro as `@mudget`, if you are ok with it, remember to **squash** and merge this PR (otherwise it will pollute the git history).

> I would directly use @bind x html"<input type=text>"

Yeah, doing this is quite convenient. But it has the missing value and multiple execution problem. @fonsp said he might be able to fix it in the future. I simply do not like red error poping up and repeated execution, even just for a second.

> I hear your love for Mustache

I like Jinja/Django templates more, but their parser is harder to implement and there is no reliable implementation in Julia.
The benefit is the parsed structure of Mustache is very easy to analyze.

#### The radio example is better
Mustache's Radio
```julia
@widget Radio """
<form id="{{groupname}}">
{{#options}}<div>{{#.}}
<input type="radio" id="{{groupname}}{{first}}" name="{{groupname}}" value="{{first}}">{{second}}
{{/.}}</div>{{/options}}
</form>
<script>
    const form = this.querySelector('#{{groupname}}')

    form.oninput = (e) => {
        form.value = e.target.value
        // and bubble upwards
    }

    // set initial value:
    const selected_radio = form.querySelector('input[checked]')
    if(selected_radio != null){
        form.value = selected_radio.value
    }
</script>
""" groupname = randstring('a':'z') default=""
```

Current Radio
```julia
struct Radio
    options::Array{Pair{<:AbstractString,<:Any},1}
    default::Union{Missing, AbstractString}
end
Radio(options::Array{<:AbstractString,1}; default=missing) = Radio([o => o for o in options], default)
Radio(options::Array{<:Pair{<:AbstractString,<:Any},1}; default=missing) = Radio(options, default)

function show(io::IO, ::MIME"text/html", radio::Radio)
    groupname = randstring('a':'z')
    withtag(io, :form, :id=>groupname) do
        for o in radio.options
            withtag(io, :div) do
                print(io, """<input type="radio" id="$(htmlesc(groupname * o.first))" name="$(groupname)" value="$(htmlesc(o.first))"$(radio.default === o.first ? " checked" : "")>""")

                withtag(io, :label, :for=>(groupname * o.first)) do
                    if showable(MIME"text/html"(), o.second)
                        show(io, MIME"text/html"(), o.second)
                    else
                        print(io, o.second)
                    end
                end
            end
        end
    end
    withtag(io, :script) do
        print(io, """
        const form = this.querySelector('#$(groupname)')

        form.oninput = (e) => {
            form.value = e.target.value
            // and bubble upwards
        }

        // set initial value:
        const selected_radio = form.querySelector('input[checked]')
        if(selected_radio != null){
            form.value = selected_radio.value
        }
        """)
    end
end

get(radio::Radio) = radio.default
```
We can not assume a people don't know Julia at all can write the second version.

**MOST IMPORTANTLY:** The first version does not need tests. There is not many people test the templates (template failures does not break the application). But the second, you must write tests.

**Comment by pankgeorg** [comment]
> We can not assume a people don't know Julia at all can write the second version.

But people are here to learn julia.

My opinion is that both versions are bad. You're writing a _language_ within ``` strings ``` and that's bad, because of (lack of) linters (are there any in this project?), lack of syntax highlighting and lack of escaping. A better solution is [this](https://github.com/fonsp/PlutoUI.jl/issues/59) but still doesn't solve the javascript part, which is also an elephant in this room.

The clock example is structured better (probably some good copy-paste occured) -- That plus some code structure plus Hyperscript should do the trick in my opinion.

**Comment by GiggleLiu** [comment]
> But people are here to learn julia.

That's fair, this change might make it harder to develop for Julia programers.
You convinced me that `Pluto` + `Mustache` should be a sperate package. Becaues it assumes someone can write template language, but is not so familar with Julia. He can use this tool to create fancy widgets.
In the future, I might consider using XML to specify widgets, so that highlight would not be an issue.

```xml
```

The problem of Hyperscript is, one needs to know both Julia and HTML well.

**Comment by fonsp** [comment]
Lots has been said already, but my two main points are:

### Mustache is an extra skill to learn
To someone who knows Mustache, this is a charming rewrite, but to someone who doesn't, this is intimidating. We should assume that most new contributors do not know Mustache, or any other templating library.

### Testing
Using Mustache does not mean that we don't need to test. I think I need to clarify what I mean with testing:

The common errors are not that we make syntax errors, and we don't forget to write `Base.get` and `Base.show` methods. The common error is in the HTML output, where I sometimes forget closing tags, forget HTML escaping and spacing. Neither the current implementation, nor this rewrite solve this problem. Mustache is not a HTML templating library, so you still make these errors. #59 would solve this.

## So.. Hyperscript?

Hyperscript is nice because it does HTML escaping + spacing + formatting correctly (and it gave me a nice feeling), but it also means that new contributors have an extra skill to learn. So also not ideal!

But why do all of this testing, if you can just look at the PlutoUI elements in a notebook and check if they work? They are simple enough for this to be possible.

## Pluto notebooks
This is why we were thinking that PlutoUI.jl should be rewritten as a series of **Pluto notebooks**, containing both the definitions, the documentation and tests. Pluto notebooks are Julia files, so the only thing needed to turn it into a package is a wrapper file that `include("the notebook files.jl")` inside a `module`. 

This is something I have been wanting to experiment with, and Pluto is the perfect package to do it with!

**Comment by GiggleLiu** [comment]
@fonsp Thanks for your comments. I agree with most of the points, but I have something to add about tests.

0. forget closing tags, forget HTML escaping and spacing
Right. But I want to emphasis intuitiveness is also good to avoiding such errors. And these frontend errors are non-breaking errors.

There are some more dangerous types of errors
1. the documentation mismatch with interface.
This is caused by repeated information input in PlutoUI. If the variables are intuitive and logicless, the docstrings can be automatically generated.

2. mismatch between the argument names and the variable names used in generating the html.
Again, cause by repeated input of variable names.

The above two errors happened again and again when I wrote the html attribute PR #39 and I have to write doctests.
The fundamental reason is the same variable was repeatedly input at least 3 times (docstring, interface, body), which violates the [DRY (don't repeat yourself)](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) rule. In the Mustache version, a variable only appear 1-2 times.

**Comment by fonsp** [comment]
About 2: this is also a shift in what I want to achieve with PlutoUI, mainly because of #52. As dpsanders suggested, people who use PlutoUI _do not care about the HTML5 spec_ - they just want a nice experience. This is also why I chose the name `default` instead of `value`.

We also talked about this on the call, but I think two good developments might be to 
1. rewrite to Pluto notebooks, and to 
2. rewrite to Hyperscript

I am unsure about both, but since this discussion is kind of stuck in speculation, I think the best thing to do is that I write both PRs, probably 1 first, 2 later, and then we can compare side-to-side.


Also, we should really write a sample notebook with more "pro" tips for writing HTML, embedded JS, custom `@bind`, etc. inside a notebook. Maybe a Mustache demo could go in there, as an alternative way of writing HTML directly.

**Comment by fonsp** [comment]
Now that https://github.com/MechanicalRabbit/HypertextLiteral.jl has been developed, we will be able to rewrite PlutoUI to use an HTML-aware templating library, awesome! @GiggleLiu If you would like to get started then that's great, if not I'll take a shot at it in about a month.

**Comment by GiggleLiu** [comment]
Nice, the package looks neat. I am a bit occupied to the talks at this moment 😭 , feel free to go ahead.

**Comment by fonsp** [comment]
Thanks for opening the discussion @GiggleLiu ! Since then, HypertextLiteral.jl was created and is very mature by now, so I started rewriting PlutoUI to use it: #139

**Pull request closed without merging**